### PR TITLE
Howto move from CRUD to Event Sourcing

### PR DESCRIPTION
**Pull request by ignasi35**
References #141

### CODE REVIEW

File: docs-source/docs/modules/concepts/pages/commands-and-events.adoc [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+
+include::partial$include.adoc[]
+
+ifdef::todo[TODO: this page is WIP]
+
+
+TODO: orders to change state VS consequences of state changes
+
+== Event Storming
+[#event_storming]
```
Comment (patriknw) [inline docs-source/docs/modules/concepts/pages/commands-and-events.adoc:14]: what is that `[#event_storming]`? (just curious)

File: docs-source/docs/modules/how-to/examples/shopping-cart-service-scala/src/main/scala/shopping/cart/ShoppingCart.scala [hunk 2]
```diff
@@ -10,11 +10,18 @@ import akka.cluster.sharding.typed.scaladsl.EntityTypeKey
 
 object ShoppingCart {
 
+  // tag::howto-crud-to-es-initial[]
   sealed trait Command
+  final case class AddItem(itemId: String, quantity: Int) extends Command
+  final case class Checkout() extends Command
```
Comment (patriknw) [inline docs-source/docs/modules/how-to/examples/shopping-cart-service-scala/src/main/scala/shopping/cart/ShoppingCart.scala:16]: `case object Checkout`

File: docs-source/docs/modules/how-to/pages/from-crud-to-eventsourcing.adoc [hunk 3]
```diff
@@ -0,0 +1,85 @@
+= Howto migrate from CRUD to Event Sourcing
+:toc:
+:toc-title: ON THIS PAGE
+:toclevels: 2
+
+include::partial$include.adoc[]
+
+When developing using a traditional CRUD approach, the entities in memory are a collection of mutable classes. Similarly, the representation of that data on the database is a collection of mutable rows in several tables.
+
+In xref:concepts:event-sourcing.adoc[Event Sourcing] (ES), we will model data as a sequence of immutable events where each event is the necessary data to mutate the data we modelled. Therefore, instead of storing the current representation of a state like we would do on a traditional CRUD approach, in Event Sourcing we store an always increasing list of events.
```
Comment (patriknw) [inline docs-source/docs/modules/how-to/pages/from-crud-to-eventsourcing.adoc:10]: do we need to introduce an abbreviation ES for Event Sourcing?

Comment (patriknw) [inline docs-source/docs/modules/how-to/pages/from-crud-to-eventsourcing.adoc:10]: s/data we modelled/state we modelled/ ?

File: docs-source/docs/modules/how-to/pages/from-crud-to-eventsourcing.adoc [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+
+[NOTE]
+====
+This howto is not about {reactive-microservices-ebook}[slicing the monolyth into microservices] {reactive-microservices-ebook}[{tab-icon}, window="tab"]
+or about the {lagom}/scala/ESAdvantage.html[Advantages of Event Sourcing] {lagom}/scala/ESAdvantage.html[{tab-icon}, window="tab"].
+====
+
+== CRUD State vs Event Sourcing State/Command/Event
+
+In CRUD, your only modelling concern is the `State` representing your business. In Event Sourcing (also referred to as `ES`), you will have to model the `State`, the `Command`'s demanding a state mutation  and the `Events`.
```
Comment (patriknw) [inline docs-source/docs/modules/how-to/pages/from-crud-to-eventsourcing.adoc:22]: ES needed?

File: docs-source/docs/modules/how-to/pages/from-crud-to-eventsourcing.adoc [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+  // fails if the cart is already checked out, or the quantity in the item field is
+  // negative or zero, or the item is on the cart already
+  def addItem(shoppingCartId, item): Unit
+
+  // fails if the cart is empty or already checked out
+  def checkOut(shoppingCartId): Unit
+}
+----
+
+In Event Sourcing, on the other hand, there is a single table (referred to as _Journal_) where we store a collection of events. Then, in code, we have a `ShoppingCart` class. We also have a function that receives a `ShoppinCartCommand` and return a new instance of `ShoppingCart` and zero, one or many `ShoppingCartEvent` 's. Finally, we have a function that given a `ShoppingCart`, and a `ShoppingCartEvent` produces the new, mutated `ShoppingCart`. Presented in pseudo-code:
```
Comment (patriknw) [inline docs-source/docs/modules/how-to/pages/from-crud-to-eventsourcing.adoc:46]: the command doesn't return "return a new instance of `ShoppingCart`"

File: docs-source/docs/modules/how-to/pages/from-crud-to-eventsourcing.adoc [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+
+1. validate the command can be processed given the current state
+2. produce the event(s) that represent the change demanded by the command given the current state
+
+The code to handle an event, on the other hand, only need to implement one step:
+
+1. produce a new state given the current state and an event
+
+See the xref:shopping-microservices:entity.adoc[step adding an Event Sourced entity]), and the xref:shopping-microservices:complete-entity.adoc[step completing the Event Sourced entity]) one on the tutorial for more details.
+
```
Comment (patriknw) [inline docs-source/docs/modules/how-to/pages/from-crud-to-eventsourcing.adoc:71]: Maybe mention how the state is recovered by replaying the events when the entity is started.

File: docs-source/docs/modules/how-to/pages/from-crud-to-eventsourcing.adoc [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+When developing using a traditional CRUD approach, the entities in memory are a collection of mutable classes. Similarly, the representation of that data on the database is a collection of mutable rows in several tables.
+
+In xref:concepts:event-sourcing.adoc[Event Sourcing] (ES), we will model data as a sequence of immutable events where each event is the necessary data to mutate the data we modelled. Therefore, instead of storing the current representation of a state like we would do on a traditional CRUD approach, in Event Sourcing we store an always increasing list of events.
+
+The first step is to enrich your business model identifying what causes a change on the state and modelling this increment.
+
+[NOTE]
+====
+This howto is not about {reactive-microservices-ebook}[slicing the monolyth into microservices] {reactive-microservices-ebook}[{tab-icon}, window="tab"]
+or about the {lagom}/scala/ESAdvantage.html[Advantages of Event Sourcing] {lagom}/scala/ESAdvantage.html[{tab-icon}, window="tab"].
```
Comment (patriknw) [inline docs-source/docs/modules/how-to/pages/from-crud-to-eventsourcing.adoc:17]: create a concept page here instead?

File: docs-source/docs/modules/how-to/pages/from-crud-to-eventsourcing.adoc [hunk 8]
```diff
@@ -0,0 +1,85 @@
...
+1. validate the command can be processed given the current state
+2. produce the event(s) that represent the change demanded by the command given the current state
+
+The code to handle an event, on the other hand, only need to implement one step:
+
+1. produce a new state given the current state and an event
+
+See the xref:shopping-microservices:entity.adoc[step adding an Event Sourced entity]), and the xref:shopping-microservices:complete-entity.adoc[step completing the Event Sourced entity]) one on the tutorial for more details.
+
+
```
Comment (patriknw) [inline docs-source/docs/modules/how-to/pages/from-crud-to-eventsourcing.adoc:72]: here is something from the Lagom docs that maybe can be incorporated in some shape?

> If you are familiar with JPA it is worth noting that a PersistentEntity can be used for similar things as a JPA `@Entity` but several aspects are rather different. For example, a JPA `@Entity` is loaded from the database from wherever it is needed, i.e. there may be many Java object instances with the same entity identifier. In contrast, there is only one instance of PersistentEntity with a given identifier. With JPA you typically only store current state and the history of how the state was reached is not captured.

**Review by patriknw — COMMENTED** [review]
looking good

File: docs-source/docs/modules/concepts/pages/commands-and-events.adoc (hunk 1 above)
Comment (ignasi35) [reply docs-source/docs/modules/concepts/pages/commands-and-events.adoc:14]: That's syntax to set a stable anchor ID. If I let asciidoctor build the anchor for me given the title it becomes `#_event_storming` (with an extra `_`). Using default anchor ID also raises a `warning` in some asciidoctor tooling (it does in the IntelliJ plugin, but it doesn't on the antora CLI).

So it's an improvement in both cosmetic and maintanibility fronts.

File: docs-source/docs/modules/how-to/pages/from-crud-to-eventsourcing.adoc (hunk 3 above)
Comment (ignasi35) [reply docs-source/docs/modules/how-to/pages/from-crud-to-eventsourcing.adoc:10]: I used the abbreviation in more places than I finally committed. I stepped back from using it and it is now only used once, os it's better if I remove the abbreviation altogether. Good catch.

File: docs-source/docs/modules/how-to/pages/from-crud-to-eventsourcing.adoc (hunk 6 above)
Comment (ignasi35) [reply docs-source/docs/modules/how-to/pages/from-crud-to-eventsourcing.adoc:71]: I'll see how to make it fit. As I was writing this page I didn't want it to become a detailed explanation of Event Sourcing and tried to put CRUD in front of Event Sourcing for comparison. I kept adding and removing bits. 
This is a relevant detail worth mentioning.

File: docs-source/docs/modules/how-to/pages/from-crud-to-eventsourcing.adoc [hunk 9]
```diff
@@ -0,0 +1,98 @@
...
+Event Sourcing applications, on the other hand, need to reconstruct the state in memory, so they can't base the state consistency on database locking. A solution is to keep a singleton memory image of the state (e.g. using {akka}/typed/cluster-sharding.html[Akka Cluster Sharding] {akka}/typed/cluster-sharding.html[{tab-icon}, window="tab"]) so all writes are redirected to a single state instance which is protected by a lock so only one modification can happen simultaneously.
+
+If you prefer Availability to Consistency, you may also use converging algorithms so concurrent modifications on multiple state instances in memory converge into a single, valid state.
+
+ifdef::todo[TODO: link to Akka's Replicated Event Sourcing docs]
```
Comment (patriknw) [inline docs-source/docs/modules/how-to/pages/from-crud-to-eventsourcing.adoc:97]: the link will be public in a few hours so you can include it now: https://doc.akka.io/docs/akka/current/typed/replicated-eventsourcing.html

**Review by patriknw — APPROVED** [review]
LGTM, with link suggestion

File: docs-source/docs/modules/how-to/pages/from-crud-to-eventsourcing.adoc (hunk 9 above)
Comment (ignasi35) [reply docs-source/docs/modules/how-to/pages/from-crud-to-eventsourcing.adoc:97]: Perfect timing!

**Review by ennru — APPROVED** [review]
This is great.

**Pull request merged**